Gems/ROS2/Code/Include/ROS2/Sensor/Events/EventSourceAdapter.h

```diff
@@ -0,0 +1,190 @@
+                        ROS2Interface::Get()->GetSimulationClock().GetExpectedSimulationLoopTime();
+
+                    if (IsPublicationDeadline(expectedLoopTime.count()))
+                    {
+                        m_sensorAdaptedEvent.Signal(args...);
+                    }
+                });
+            m_eventSource.ConnectToSourceEvent(m_adaptedEventHandler);
+            m_eventSource.Activate();
+        }
+
+        //! Deactivates event source adapter - deactivates event source and disconnects internal adapted event handler from source event. If
+        //! it will be necessary, this implementation allows multiple consecutive calls of Activate / Deactivate, however user must also
+        //! investigate specific event source implementation with such case in mind.
+        void Deactivate()
+        {
+            m_eventSource.Deactivate();
+            m_adaptedEventHandler.Disconnect();
+        }
+
+        //! Parses sensor configuration into event source configuration.
+        //! @param sensorConfiguration Configuration of sensor using this event source adapter.
+        void Configure(const SensorConfiguration& sensorConfiguration)
+        {
+            m_eventSource.Configure(sensorConfiguration);
+            m_adaptedFrequency = sensorConfiguration.m_frequency;
+        }
+
+        //! Connects given event handler to source event (ROS2::SensorEventSource). That event is signalled regardless of adapted frequency
+        //! set for event source adapter (ROS2::EventSourceAdapter::Configure). Its frequency depends only on specific event source
+        //! implementation. If specific working frequency is required (main purpose of ROS2::EventSourceAdapter), user should see
+        //! ROS2::EventSourceAdapter::ConnectToAdaptedEvent method.
+        //! @param sourceEventHandler Event handler for source event (frequency not managed by event source adapter).
+        //! @see ROS2::SensorEventSource
+        void ConnectToSourceEvent(typename EventSourceT::SourceEventHandlerType& sourceEventHandler)
+        {
+            m_eventSource.ConnectToSourceEvent(sourceEventHandler);
+        }
+
+        //! Connects given event handler to adapted event (ROS2::EventSourceAdapter). This event is signalled with a frequency set in event
+        //! source adapter configuration (ROS2::EventSourceAdapter::Configure).
+        //! @param adaptedEventHandler Event handler for adapted event (frequency set through event source adapter configuration).
+        void ConnectToAdaptedEvent(typename EventSourceT::SourceEventHandlerType& adaptedEventHandler)
+        {
+            adaptedEventHandler.Connect(m_sensorAdaptedEvent);
+        }
+
+    private:
+        //! Uses tick counter, expected loop time and frequency set for adapter to support managing calls from event source.
+        //! @param expectedLoopTime Expected simulation loop time.
+        //! @return Whether it is time to signal adapted event.
```

Comment:
It is feasible to want a different mechanism for approximating frequencies, but this does not need to be in this PR.